From a89b791e0650908f847746597d14efe432ead3f5 Mon Sep 17 00:00:00 2001
From: throwaway96 <68320646+throwaway96@users.noreply.github.com>
Date: Sun, 26 Feb 2023 15:29:08 -0500
Subject: [PATCH] webOS librt compatibility fix

Automatically links against librt when needed. Disable with -tno-webos-compat.
---
 gcc/config.gcc             |  7 +++++++
 gcc/config/linux-webos.h   | 15 +++++++++++++++
 gcc/config/linux-webos.opt |  6 ++++++
 3 files changed, 28 insertions(+)
 create mode 100644 gcc/config/linux-webos.h
 create mode 100644 gcc/config/linux-webos.opt

diff --git a/gcc/config.gcc b/gcc/config.gcc
index c5064dd37..90db0cf74 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -5915,3 +5915,10 @@ i[34567]86-*-* | x86_64-*-*)
 	fi
 	;;
 esac
+
+case ${target} in
+*-webos-*|*-lgtv-*)
+	tm_file="${tm_file} linux-webos.h"
+	extra_options="${extra_options} linux-webos.opt"
+	;;
+esac
diff --git a/gcc/config/linux-webos.h b/gcc/config/linux-webos.h
new file mode 100644
index 000000000..8575ea893
--- /dev/null
+++ b/gcc/config/linux-webos.h
@@ -0,0 +1,15 @@
+#ifdef HAVE_LD_AS_NEEDED
+# define WEBOS_AS_NEEDED(_LIB) LD_AS_NEEDED_OPTION " " _LIB " " LD_NO_AS_NEEDED_OPTION
+#else
+# define WEBOS_AS_NEEDED(_LIB) _LIB
+#endif
+
+#undef GNU_USER_TARGET_NO_PTHREADS_LIB_SPEC
+#define GNU_USER_TARGET_NO_PTHREADS_LIB_SPEC \
+  "%{!tno-webos-compat:" \
+    "%{shared:-lc " WEBOS_AS_NEEDED("-lrt") "} " \
+    "%{!shared:%{profile:-lc_p " WEBOS_AS_NEEDED("-lrt_p") "}%{!profile:-lc " WEBOS_AS_NEEDED("-lrt") "}}" \
+  ";:" \
+    "%{shared:-lc} " \
+    "%{!shared:%{profile:-lc_p}%{!profile:-lc}}" \
+  "}"
diff --git a/gcc/config/linux-webos.opt b/gcc/config/linux-webos.opt
new file mode 100644
index 000000000..0357773d4
--- /dev/null
+++ b/gcc/config/linux-webos.opt
@@ -0,0 +1,6 @@
+; webOS specific options.
+
+tno-webos-compat
+Driver
+Disable all webOS compatibility hacks.
+
-- 
2.39.2

